packages/webview_flutter/webview_flutter_android/android/src/main/java/io/flutter/plugins/webviewflutter/WebViewHostApiImpl.java

```diff
@@ -85,6 +171,76 @@ public void onInputConnectionLocked() {
     public void onInputConnectionUnlocked() {
       unlockInputConnection();
     }
+
+    @Override
+    public void setWebViewClient(WebViewClient webViewClient) {
+      super.setWebViewClient(webViewClient);
+      if (currentWebViewClient instanceof Releasable) {
+        ((Releasable) currentWebViewClient).release();
+      }
+      currentWebViewClient = (WebViewClient) webViewClient;
```

Comment:
Does Java have the ability to do subclass-restricted templates? You have this repeated pattern; it would be nice if you could instead have a simple smart pointer class that can only be created for things that inherit from Releasable (but is still typed with the correct specific type), that would handle the call-release-if-necessary-then-assign logic internally.

Reply:
I wasn't sure what you meant by this. But, my assumption is that you were thinking something like this

```java
private static class ReleasableChild<T> {
    @Nullable private T child;

    private ReleasableChild(@Nullable T child) {
      this.child = child;
    }

    private void release() {
      if (child instanceof Releasable) {
        ((Releasable) child).release();
      }
      child = null;
    }
  }
```

I replaced all the variables with this value.

Reply:
(It looks like the Java name for what I was looking for support for is "bounded types".)

This still relies on `release` being called manually on every assignment, which is error-prone. I was looking for something like (probably doesn't compile as written since it's freehand):
```
  private static class ReleasableValue<T extends Releasable> {
    @Nullable private T value;

    ReleasableValue(@Nullable T value) {
      this.value = value;
    }

    void set(@Nullable T newValue) {
      release();
      value = newValue;
    }

    @Nullable T get() {
      return value;
    }

    void release() {
      if (value != null) {
        value.release();
      }
      value = null;
    }

    void finalize() {
      release();
    }
  }
```

Reply:
ah ok that makes sense. I changed it to:

```java
private static class ReleasableValue<T> {
    @Nullable private T value;

    ReleasableValue() {}

    ReleasableValue(@Nullable T value) {
      this.value = value;
    }

    void set(@Nullable T newValue) {
      release();
      value = newValue;
    }

    @Nullable
    T get() {
      return value;
    }

    void release() {
      if (value instanceof Releasable) {
        ((Releasable) value).release();
      }
      value = null;
    }
  }
```

I removed `finalize` since it was only calling `release`. I also changed `ReleasableValue<T extends Releasable>` to `ReleasableValue<T>` because the `WebView.setWebViewClient` will take either a `WebViewClientImpl` or `WebViewClientCompatImpl` and their only shared interface is `WebViewClient`; which doesn't extend `Releasable`.

Reply:
But I suppose I could just cast `WebViewClient` to `Releasable` as I set it. Maybe that is a better solution?